cpp/src/reductions/reduction.cuh

```diff
@@ -0,0 +1,126 @@
+ * @param[in] stream    cuda stream
+ * @returns   scalar    output scalar in device memory
+ *
+ * @tparam Op               the device binary operator
+ * @tparam InputIterator    the input column iterator
+ * @tparam OutputType       the output type of reduction
+ * ----------------------------------------------------------------------------**/
+template <typename Op, typename InputIterator, typename OutputType=typename thrust::iterator_value<InputIterator>::type>
+std::unique_ptr<scalar> reduce(InputIterator d_in, cudf::size_type num_items, Op op,
+  rmm::mr::device_memory_resource* mr, cudaStream_t stream)
+{
+  typename Op::Op binary_op{};
+  OutputType identity = Op::Op::template identity<OutputType>();
+  rmm::device_scalar<OutputType> dev_result{identity, stream, mr};
+
+  // Allocate temporary storage
+  rmm::device_buffer d_temp_storage;
+  size_t temp_storage_bytes = 0;
+  cub::DeviceReduce::Reduce(d_temp_storage.data(), temp_storage_bytes, d_in, dev_result.data(), num_items, binary_op, identity, stream);
+  d_temp_storage = rmm::device_buffer{temp_storage_bytes, stream, mr};
+
+  // Run reduction
+  cub::DeviceReduce::Reduce(d_temp_storage.data(), temp_storage_bytes, d_in, dev_result.data(), num_items, binary_op, identity, stream);
+
+  using ScalarType = cudf::experimental::scalar_type_t<OutputType>;
+  auto s = new ScalarType(std::move(dev_result), true, stream, mr); //only for string_view, data is copied
+  return std::unique_ptr<scalar>(s);
+}
+
+/** --------------------------------------------------------------------------*
+ * @brief compute reduction by the compound operator (reduce and transform)
+ *
+ * @param[in] d_in      the begin iterator
+ * @param[in] num_items the number of items
+ * @param[in] op        the reduction operator 
+ * @param[in] valid_count   the intermediate operator argument 1
+ * @param[in] ddof      the intermediate operator argument 2
+ * @param[in] stream    cuda stream
+ * @returns   scalar    output scalar in device memory
+ *
+ * The reduction operator must have intermediate::compute_result() method.
+ * This method performs reduction using binary operator `Op::Op` and transforms the
+ * result to `OutputType` using `compute_result()` transform method.
+ *
+ * @tparam Op               the reduction operator
+ * @tparam InputIterator    the input column iterator
+ * @tparam OutputType       the output type of reduction
+ * ----------------------------------------------------------------------------**/
+template <typename Op, typename InputIterator, typename OutputType, 
+  typename IntermediateType=typename thrust::iterator_value<InputIterator>::type>
+  //FIXME: adding this causes reduction.cuh:118:401:   error : ‘__T290’ was not declared in this scope
+  //typename std::enable_if_t< std::is_base_of<cudf::experimental::reduction::op::CompoundOp<Op>, Op>::value, int> = 0>
+std::unique_ptr<scalar> reduce(InputIterator d_in, cudf::size_type num_items, Op op, 
```

Comment:
Since CRTP is being used, this enforces the interface w/o needing an `enable_if`. 
```suggestion
std::unique_ptr<scalar> reduce(InputIterator d_in, cudf::size_type num_items, CompoundOp<Op> op, 
```